### ISSUE TITLE
New Operator: RegexFullMatch

### ISSUE THREAD

**Issue by adityagoel4512** [onnx/onnx#5317]
# New Operator: RegexFullMatch

### Describe the operator

ONNX currently lacks support for regex matching. Regex matching is an incredibly useful string operation that is pervasive in feature preprocessing. Given an input tensor of type `string`, the operator would return a `boolean` array denoting whether a full match has been achieved. The output tensor would be the same shape as the input as matching is done elementwise.

In Tensorflow this is directly supported via https://www.tensorflow.org/api_docs/python/tf/strings/regex_full_match which specifically supports regex patterns based on re2 syntax https://github.com/google/re2/wiki/Syntax. The behaviour of this operator should reflect the examples in the link.

Attributes:
-  Given the lack of complete standardisation of regex syntax, I would propose an attribute called `syntax` which is a literal `string` denoting the specific accepted syntax. Options can be `"ECMA"` and `"posix"`. For runtimes/backends, there is a clear implementation path for these options across different languages. 
- The regex pattern to match on can be specified as a `string` attribute called `pattern`.

### Can this operator be constructed using existing onnx operators?
No.

### Is this operator used by any model currently? Which one?
This operator is very commonly used during data preprocessing where we have non-categorical string data. It is used to generate masks for filtering, validation and replacement operations.

Commonly used DataFrame oriented libraries such as `pandas` and `polars` provide commonly used capabilities for regex based replacement ([pandas](https://pandas.pydata.org/docs/reference/api/pandas.Series.replace.html), [polars](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.str.replace.html)) and for filtering/membership checking ([pandas](https://pandas.pydata.org/docs/reference/api/pandas.Series.filter.html), [polars](https://pola-rs.github.io/polars/py-polars/html/reference/series/api/polars.Series.str.contains.html)) is present in `pandas`. Combining a `RegexFullMatch` operator with the existing operator set will unlock these very common operations without requiring numerous additional operators in the standard.

### Are you willing to contribute it? (Y/N)
Y

**Comment by xadupre** [comment]
One version of it is implemented in onnxruntime-extensions: https://github.com/microsoft/onnxruntime-extensions/blob/main/docs/custom_ops.md. Would that work for you?

**Comment by adityagoel4512** [comment]
The regex options in onnxruntime-extensions don't quite do the same thing as the proposed operator.

On a more general note, I'd strongly prefer the standardisation route for this operator since I believe it is sufficiently general purpose and will be widely used by the community. Relying on onnxruntime-extensions also assumes that users use both onnxruntime and onnxruntime-extensions. In our specific case, we don't ship with onnxruntime-extensions for a few reasons.

**Comment by WilliamTambellini** [comment]
I agree with @adityagoel4512 that we would also prefer a std onnx op for regex. Best.

**Comment by WilliamTambellini** [comment]
FYI
https://pytorch.org/text/0.15.0/transforms.html#regextokenizer

**Comment by wenbingl** [comment]
> The regex options in onnxruntime-extensions don't quite do the same thing as the proposed operator. StringECMARegexReplace seems to do a regex based replacement and StringECMARegexSplit does a string splitting operation, while this operator just does a match.
> 
> On a more general note, I'd strongly prefer the standardisation route for this operator since I believe it is sufficiently general purpose and will be widely used by the community. Relying on onnxruntime-extensions assumes that users use onnxruntime and onnxruntime-extensions which of course may not be true (e.g. onnx-mlir and related projects). In our specific case, we don't ship with onnxruntime-extensions for a few reasons.



> The regex options in onnxruntime-extensions don't quite do the same thing as the proposed operator. StringECMARegexReplace seems to do a regex based replacement and StringECMARegexSplit does a string splitting operation, while this operator just does a match.
> 
> On a more general note, I'd strongly prefer the standardisation route for this operator since I believe it is sufficiently general purpose and will be widely used by the community. Relying on onnxruntime-extensions assumes that users use onnxruntime and onnxruntime-extensions which of course may not be true (e.g. onnx-mlir and related projects). In our specific case, we don't ship with onnxruntime-extensions for a few reasons.

onnxruntime-extensions also already had google/re2 based operators here, https://github.com/microsoft/onnxruntime-extensions/tree/main/operators/text/re2_strings. And it is very convenient to add new one if the existing two doesn't meet your requirement.

Can you share more details about the reason that onnxruntime-extensions wasn't ship in your case? Not using onnxruntime for inference, Or something else?

**Comment by cbourjau** [comment]
> Can you share more details about the reason that onnxruntime-extensions wasn't ship in your case? Not using onnxruntime for inference, Or something else?

Sure! We would like to avoid `onnxruntime-extensions` for a few reasons:
- It is a fairly complex dependency for something that seems fairly small
- Error handling for fallible operators such as this one is [currently based on undefined behavior](https://github.com/microsoft/onnxruntime/issues/14287#issuecomment-1382350705) as far as I can tell. We have to be able to guarantee a safe and stable deployment of our models many years into the future and I'm not comfortable doing that on the basis of `onnxruntime-extensions` at this point.
- For the above and other reasons we would like to be as independent of any given runtime as reasonably possible.

**Comment by wenbingl** [comment]
> > Can you share more details about the reason that onnxruntime-extensions wasn't ship in your case? Not using onnxruntime for inference, Or something else?
> 
> Sure! We would like to avoid `onnxruntime-extensions` for a few reasons:
> 
> * It is a fairly complex dependency for something that seems fairly small
> * Error handling for fallible operators such as this one is [currently based on undefined behavior](https://github.com/microsoft/onnxruntime/issues/14287#issuecomment-1382350705) as far as I can tell. We have to be able to guarantee a safe and stable deployment of our models many years into the future and I'm not comfortable doing that on the basis of `onnxruntime-extensions` at this point.
> * For the above and other reasons we would like to be as independent of any given runtime as reasonably possible.

Thanks for sharing the concerns, and there are some design in ort-extensions for addressing them.
1. dependency, All operators are configurable and can be disabled as well as their dependencies, like cmake option  OCOS_ENABLE_TF_STRING.
2. Fallible KernelComputeV2 ABI is added ORT recently, and ort-extensions will soon migrate to this new ABI.
3. Operators are also backward compatible, any replacing and deleting any input/output in operator signature will trigger an upgrading in the operator definition.
